cmd/soroban-cli/src/signer/types.rs

```diff
@@ -0,0 +1,183 @@
+            signature: Signature(tx_signature.try_into()?),
+        })
+    }
+}
+
+/// A trait for signing Stellar transactions and Soroban authorization entries
+#[async_trait::async_trait]
+pub trait Transaction {
+    /// Sign a Stellar transaction with the given source account
+    /// This is a default implementation that signs the transaction hash and returns a decorated signature
+    ///
+    /// Todo: support signing the transaction directly.
+    /// # Errors
+    /// Returns an error if the source account is not found
+    async fn sign_txn(
+        &self,
+        source_account: &stellar_strkey::ed25519::PublicKey,
+        txn: &xdr::Transaction,
+        network: &Network,
+    ) -> Result<DecoratedSignature, Error>;
+}
+
+#[async_trait::async_trait]
+impl<T> Transaction for T
+where
+    T: TransactionHash + Send + Sync,
+{
+    async fn sign_txn(
+        &self,
+        source_account: &stellar_strkey::ed25519::PublicKey,
+        txn: &xdr::Transaction,
+        Network {
+            network_passphrase, ..
+        }: &Network,
+    ) -> Result<DecoratedSignature, Error> {
+        let hash = transaction_hash(txn, network_passphrase)?;
+        self.sign_txn_hash(source_account, hash).await
+    }
+}
+pub async fn sign_txn_env(
+    signer: &(impl Transaction + std::marker::Sync),
+    source_account: &stellar_strkey::ed25519::PublicKey,
+    txn_env: TransactionEnvelope,
+    network: &Network,
+) -> Result<TransactionEnvelope, Error> {
+    match txn_env {
+        TransactionEnvelope::Tx(TransactionV1Envelope { tx, signatures }) => {
+            let decorated_signature = signer.sign_txn(source_account, &tx, network).await?;
+            let mut sigs = signatures.to_vec();
+            sigs.push(decorated_signature);
+            Ok(TransactionEnvelope::Tx(TransactionV1Envelope {
+                tx,
+                signatures: sigs.try_into()?,
+            }))
+        }
+        _ => Err(Error::UnsupportedTransactionEnvelopeType),
+    }
+}
+
+pub(crate) fn hash(network_passphrase: &str) -> xdr::Hash {
+    xdr::Hash(Sha256::digest(network_passphrase.as_bytes()).into())
+}
+
+pub struct LocalKey {
+    key: ed25519_dalek::SigningKey,
+    prompt: bool,
+}
+
+impl LocalKey {
+    pub fn new(key: ed25519_dalek::SigningKey, prompt: bool) -> Self {
+        Self { key, prompt }
+    }
+}
+
+#[async_trait::async_trait]
+impl Blob for LocalKey {
+    async fn sign_blob(&self, data: &[u8]) -> Result<Vec<u8>, Error> {
+        if self.prompt {
+            eprintln!("Press 'y' or 'Y' for yes, any other key for no:");
+            match read_key() {
+                'y' | 'Y' => {
+                    eprintln!("Signing now...");
+                }
+                _ => return Err(Error::UserCancelledSigning),
+            };
+        }
+        let sig = self.key.sign(data);
+        Ok(sig.to_bytes().to_vec())
+    }
+}
+
+pub fn read_key() -> char {
```

Comment:
I noticed a potential issue with this functionality while trying to add it to the `sign_with_lab` [branch](https://github.com/AhaLabs/stellar-cli/pull/25#discussion_r1755504162). Also possible that I am missing something or calling the command incorrectly.

When I am piping my tx xdr into the cli, the program seems to not have access to the stdin reader when it gets to trying to read the user's y response. In other words, when I run this command:
```
echo AAAAAgAAAADcOHnq5sGLOngOCEMyLqqn5CvFV2HGbOSjJAIzhqBdkAAAAGQAEb7UAAAAAQAAAAAAAAAAAAAAAQAAAAAAAAAYAAAAAQAAAAAAAAAAAAAAANw4eermwYs6eA4IQzIuqqfkK8VXYcZs5KMkAjOGoF2Q04SY/PMseMCgfbEFkX5gTr9qD/d0mvD9H3acFOz/X5oAAAAALtGgcgsAibk5q8VOLY2R1G1NdGX7fLzlz7iBFBy64JIAAAAAAAAAAAAAAAA= \
| stellar tx sign --network testnet --sign-with-key testnet-alice
```

I'm getting the same error that I'm seeing over on the `sign_with_lab` PR:
```
thread 'main' panicked at cmd/soroban-cli/src/signer.rs:164:41:
called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: "Failed to initialize input reader" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
> The nice thing about the design of the tx sign UI is there is no need to ask for y. The command already contains a very clear authorization to sign. The user is literally stating --sign-with-key and so additional confirmation is not required imo.

This makes sense to me. 👍  I think this could be the case for most of (all of?) the `sign-with-*` options.

Reply:
Removed 👍